code/API_definitions/qod-api.yaml

```diff
@@ -414,14 +432,50 @@ components:
         ipv6Address:
           $ref: "#/components/schemas/Ipv6Address"
       description: Application server identifier
+
     NetworkAccessIdentifier:
       type: string
       example: "123456789@domain.com"
+
     PhoneNumber:
       type: string
       pattern: '^\+?[0-9]{5,15}$'
       example: "123456789"
-      description: Subscriber number in E.164 format (starting with country code). Optionally prefixed with '+'.
+      description: Subscriber number (MSISDN) in E.164 format, starting with country code and ptionally prefixed with '+'.
+
+    DeviceIpv4Addr:
+      type: object
+      properties:
+        publicAddress:
+          $ref: "#/components/schemas/SingleIpv4Addr"
+        privateAddress:
+          $ref: "#/components/schemas/SingleIpv4Addr"
+        publicPort:
+          $ref: "#/components/schemas/Port"
+      oneOf:
```

Comment:
This `oneOf` triggered some debate. To me is acceptable at this level as it may be used by JSON-schema validators. 
An alternative `anyOf` would allow to send all 3 properties, which is something that I'm not sure if want to prevent.
I have to further validate this also with our dev team.

Reply:
My view was that this was the best way of presenting valid options to the API consumer. It does not prevent them sending all 3 parameters if known, but does not mandate it either, as both `privateAddress` and `publicPort` could be difficult to obtain for different scenarios.

For implementation, discriminators continue to cause problems, and may do for some time. My view if that implementors should be prepared to modify the CAMARA API definition for our own tooling. A certain amount of that has to be done anyway, for `apiRoot` or `securitySchemes` for example. In Vodafone, for example, we need to change all the lowerCamelCase `operationId`s to snake_case, but that is fine as it does not affect the API consumer. Workarounds for discriminators are also possible if not natively supported by tooling.

Reply:
@jlurien 

If I understand correctly, then:
```
{
  "publicAddress": "84.125.93.10",
  "privateAddress": "10.10.10.10",
  "publicPort" : 59765
}
```
would be rejected because it matches both:
```
    DeviceIpv4Addr:
      type: object
      properties:
        publicAddress:
          $ref: "#/components/schemas/SingleIpv4Addr"
        privateAddress:
          $ref: "#/components/schemas/SingleIpv4Addr"
        publicPort:
          $ref: "#/components/schemas/Port"
      required: [publicAddress, privateAddress]
```
and
```
    DeviceIpv4Addr:
      type: object
      properties:
        publicAddress:
          $ref: "#/components/schemas/SingleIpv4Addr"
        privateAddress:
          $ref: "#/components/schemas/SingleIpv4Addr"
        publicPort:
          $ref: "#/components/schemas/Port"
      required: [publicAddress, publicPort]
```
I can see that might be true (I have no way to test these things), and can see `anyOf` is "more" unambiguous (to machines at least, if not to developers), so I'll change `oneOf` to `anyOf`.

The reason to allow all three parameters is that it is much easier to identify a device from `publicAddress` and `privateAddress` than from `publicAddress` and `publicPort`. So if the API caller knows `publicAddress`, we should encourage them to give it, even if they also know `publicPort`.

Reply:
Yes, you understood right. Some guidelines recommend to use `anyOf` in most cases instead of oneOf, as it is more efficient for validators. oneOf requires to validate each schema and assure that only one is true, while validation of anyOf can return true when first validation succeeded